src/Illuminate/Foundation/Application.php

```diff
@@ -245,6 +261,65 @@ protected function registerBaseServiceProviders()
         $this->register(new RoutingServiceProvider($this));
     }
 
+    /**
+     * Register the standard kernel classes for the application.
+     *
+     * @return $this
+     */
+    protected function withKernels()
+    {
+        $this->singleton(
+            \Illuminate\Contracts\Http\Kernel::class,
+            \App\Http\Kernel::class
+        );
+
+        $this->singleton(
+            \Illuminate\Contracts\Console\Kernel::class,
+            \App\Console\Kernel::class
+        );
+
+        return $this;
+    }
```

Comment:
It's relatively common in DDD-based and other software architecture to move the `Kernel` classes to other namespaces. Can we make this more easily configurable?

```suggestion
    public function withKernels(?string $httpKernel = null, string $consoleKernel = null)
    {
        $this->singleton(
            \Illuminate\Contracts\Http\Kernel::class,
            $httpKernel ?? \App\Http\Kernel::class
        );

        $this->singleton(
            \Illuminate\Contracts\Console\Kernel::class,
            $consoleKernel ?? \App\Console\Kernel::class
        );

        return $this;
    }
```

Reply:
might just have it as `$http` and `$console` so we can do `withKernels(console: Acme\Console\Kernel::class)`

Reply:
After a second thought, one can rebind the kernel singletons before the application is actually run. `./bootstrap/app.php` just builds the application object but doesn't run it yet.

From the code on the companion PR on `laravel/laravel` you could change the return statement to something like this:

```php
$app = Application::create()
    ->withBroadcasting()
    ->withExceptionHandling(function ($handler) {
    });

$app->singleton(
    Illuminate\Contracts\Http\Kernel::class,
    App\HttpKernel::class
);

$app->singleton(
    Illuminate\Contracts\Console\Kernel::class,
    App\ConsoleKernel::class
);

$app->singleton(
    Illuminate\Contracts\Debug\ExceptionHandler::class,
    App\ExceptionHandler::class
);

return $app;
```

And retain the previous ability to customize the kernels and exception handler namespaces.

Reply:
I think I would be okay with that. 👍